File: src/c/src/idtp.c
// SPDX-License-Identifier: Apache-2.0.
// Copyright (C) 2025-present idtp project and contributors.

//! Inertial Measurement Unit Data Transfer Protocol frame implementation.

#include <idtp/idtp.h>
#include <string.h>

IdtpFrame idtp_frame_create(void) {
    const IdtpFrame frame = (IdtpFrame) {
        .header = idtp_header_create(),
        .payload = {0},
        .payload_size = 0,
    };

    return frame;
}

void idtp_frame_set_header(IdtpFrame *self, const IdtpHeader *header) {
    self->header = *header;
}

void idtp_frame_set_payload(IdtpFrame *self, const uint8_t *payload, const size_t size) {
    if (size <= IDTP_PAYLOAD_MAX_SIZE) {
        memcpy(self->payload, payload, size);
        self->payload_size = size;
    }
}

void idtp_frame_pack(const IdtpFrame *self, uint8_t *buffer) {
    size_t lower_range = 0;

    IdtpHeader header = self->header;
    idtp_header_convert_endian(&header);

    memcpy(&buffer[lower_range], &header, IDTP_HEADER_SIZE);
    lower_range += IDTP_HEADER_SIZE;

    memcpy(&buffer[lower_range], self->payload, self->payload_size);
    lower_range += self->payload_size;

    memcpy(&buffer[lower_range], &IDTP_TRAILER, IDTP_TRAILER_SIZE);
}

IdtpFrame idtp_frame_from_bytes(const uint8_t *bytes, const size_t size) {
    IdtpFrame frame = {0};
    frame.header = idtp_header_from_bytes(bytes);
    frame.payload_size = size - IDTP_PACKET_MIN_SIZE;
    memcpy(frame.payload, &bytes[IDTP_HEADER_SIZE], frame.payload_size);

    return frame;
}

File: src/rust/idtp/src/idtp.rs
// SPDX-License-Identifier: Apache-2.0.
// Copyright (C) 2025-present idtp project and contributors.

//! Inertial Measurement Unit Data Transfer Protocol frame implementation.

use crate::{IDTP_HEADER_SIZE, IDTP_TRAILER, IDTP_TRAILER_SIZE, IdtpHeader};

/// IDTP network packet max size in bytes. It includes size of IDTP header,
/// payload and packet trailer.
pub const IDTP_PACKET_MAX_SIZE: usize = 1024;

/// IDTP network packet min size in bytes.
pub const IDTP_PACKET_MIN_SIZE: usize = IDTP_HEADER_SIZE + IDTP_TRAILER_SIZE;

/// IDTP network packet payload max size in bytes.
pub const IDTP_PAYLOAD_MAX_SIZE: usize =
    IDTP_PACKET_MAX_SIZE - IDTP_HEADER_SIZE - IDTP_TRAILER_SIZE;

/// Inertial Measurement Unit Data Transfer Protocol frame struct.
#[derive(Debug, Clone, Copy)]
pub struct IdtpFrame {
    /// IDTP network packet header.
    header: IdtpHeader,
    /// Value that containing IMU data.
    payload: [u8; IDTP_PAYLOAD_MAX_SIZE],
    /// IDTP payload size in bytes.
    payload_size: usize,
}

impl IdtpFrame {
    /// Construct new `IdtpFrame` struct.
    ///
    /// # Returns
    /// - New `IdtpFrame` struct.
    pub fn new() -> Self {
        Self {
            header: IdtpHeader::new(),
            payload: [0u8; IDTP_PAYLOAD_MAX_SIZE],
            payload_size: 0usize,
        }
    }

    /// Set IDTP header.
    ///
    /// # Parameters
    /// - `header` - given IDTP header to set.
    pub fn set_header(&mut self, header: &IdtpHeader) {
        self.header = *header;
    }

    /// Set IDTP payload.
    ///
    /// # Parameters
    /// - `payload` - given IDTP payload bytes to set.
    pub fn set_payload(&mut self, payload: &[u8]) {
        let payload_size = payload.len();

        if payload_size <= IDTP_PAYLOAD_MAX_SIZE {
            self.payload[0..payload_size].copy_from_slice(payload);
            self.payload_size = payload_size;
        }
    }

    /// Get IDTP header.
    ///
    /// # Returns
    /// - IDTP header struct.
    pub fn header(&self) -> IdtpHeader {
        self.header
    }

    /// Get IDTP payload.
    ///
    /// # Returns
    /// - IDTP payload in bytes representation.
    pub fn payload(&self) -> &[u8] {
        &self.payload[0..self.payload_size]
    }

    /// Get IDTP payload size in bytes.
    ///
    /// # Returns
    /// - IDTP payload in bytes representation.
    pub fn payload_size(&self) -> usize {
        self.payload_size
    }

    /// Pack into raw IDTP network packet.
    ///
    /// # Parameters
    /// - `buffer` - given buffer to store raw IDTP packet.
    ///
    /// # Returns
    /// - `Ok`  - in case of success.
    /// - `Err` - otherwise.
    ///
    /// # Errors
    /// - Will return `Err` if network packet buffer size is too small.
    pub fn pack(&self, buffer: &mut [u8]) -> Result<(), &'static str> {
        if buffer.len() < IDTP_PACKET_MIN_SIZE + self.payload_size {
            return Err("Network packet buffer size too small");
        }

        let mut lower_range = 0;
        let mut upper_range = IDTP_HEADER_SIZE;

        buffer[lower_range..upper_range]
            .copy_from_slice(&self.header.as_bytes_be());
        lower_range = upper_range;
        upper_range += self.payload_size;

        buffer[lower_range..upper_range]
            .copy_from_slice(&self.payload[0..self.payload_size]);
        lower_range = upper_range;
        upper_range += IDTP_TRAILER_SIZE;

        buffer[lower_range..upper_range].copy_from_slice(IDTP_TRAILER);

        Ok(())
    }
}

impl Default for IdtpFrame {
    /// Construct new default `IdtpFrame` struct.
    ///
    /// # Returns
    /// - New default `IdtpFrame` struct.
    fn default() -> Self {
        Self::new()
    }
}

impl From<&[u8]> for IdtpFrame {
    /// Convert byte slice to IDTP frame.
    ///
    /// # Parameters
    /// - `bytes` - given byte slice to convert (big-endian byte order).
    ///
    /// # Returns
    /// - IDTP frame struct from byte slice.
    fn from(bytes: &[u8]) -> Self {
        let mut idtp = IdtpFrame::new();
        idtp.header = IdtpHeader::from(&bytes[0..IDTP_HEADER_SIZE]);
        idtp.payload_size = bytes.len() - IDTP_HEADER_SIZE - IDTP_TRAILER_SIZE;

        let copy_range = IDTP_HEADER_SIZE..IDTP_HEADER_SIZE + idtp.payload_size;
        idtp.payload[0..idtp.payload_size].copy_from_slice(&bytes[copy_range]);
        idtp
    }
}
